[miio] Add a few new vacuum robots and station channels

truidix opened the pull request:
I added some new channels and also capabilities.
- optional channel for base status (e.g. dirt water tank full)
- controling cleaning the mop
- check if mop drying is enabled
- remaining time for mop drying

marcelrv commented:
@truidix unfortunately there was a PR pending that caused a conflict, can you resolve that one.
Also I did a quick glance at the code, I was struggling to see how some of these optional channels get enabled.

After the merge conflict is resolved will take deeper look

truidix commented:
Hey @marcelrv ,

that's strange.  I synchronized may branch before merging.  Anyways I tried to merge it from my mobile. I can't see what's wrong.


About the ne Channels: I did it like you didn't before: checking at the configuration if there is a specific tag available and then enabling the new channels. I'm having two S5 and one Q Revo, so I could test both.

jlaur commented:
@truidix - can you also apply spotless so that the build will hopefully succeed?

truidix commented:
Hey @jlaur,

sry i was away for a few days. i dig the merge in github, i iwll have a look into this in the next few days.

truidix commented:
Hey @jlaur ,

I appley spotless, and fixed a compilation (copy&paste) error.. :-)


So, we are good to go?

marcelrv commented:
@truidix 

Appreciate your PR!

2 more comments:
1) The label in the xml and in the i18 files are not in sync. Can you make these in sync, meaning that the label you have in English in the channel XML is the same as in the English's translation i18 file.

2) Can you please also add the new channels to the readme
Note you need to modify the readme base file, otherwise the change is automatically overwritten by the next update 
https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.miio/README.base.md#robo-rock-vacuum-channels

truidix commented:
@marcelrv , sure I added changed the labels and added channels to the readme.base

marcelrv commented on bundles/org.openhab.binding.miio/README.base.md:
```diff
@@ -240,6 +240,8 @@ e.g. `openhab:send actionCommand 'upd_timer["1498595904821", "on"]'` would enabl
 | String  | info#fw_features                  | Firmware Features          |
 | String  | info#room_mapping                 | Room Mapping details       |
 | String  | info#multi_maps_list              | Maps Listing details       |
+| Number  | status#mop_drying_time            | Mop Drying Time            |
+| Switch  | status#is_mop_drying              | Mop cleaning active        |
```
suggest to move these also to the section about the additional depending on the capabilities, as these are defiantly not available on all robots

truidix replied:
I misses that. 

I moved them underneath the additional capabilities

marcelrv commented:
LGTM.
@jlaur @wborn  can one of you review it for merging.as well. Thanks!

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -299,6 +315,18 @@
 		<item-type>String</item-type>
 		<label>Vacuum Room [room#]</label>
 	</channel-type>
+	<channel-type id="collectdust" advanced="true">
```
Shouldn't this be:
```suggestion
	<channel-type id="collect_dust" advanced="true">
```

for staying consistent with the binding naming convention?

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -299,6 +315,18 @@
 	</channel-type>
+	<channel-type id="collectdust" advanced="true">
+		<item-type>Switch</item-type>
+		<label>Start collect the dust at the station</label>
```
See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions:
```suggestion
		<label>Start Collect the Dust at the Station</label>
```

however, it should also be kept short (2-3 words with up to 25 chars). Perhaps you could shorten the label and provide the current label as `description` instead?

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -299,6 +315,18 @@
+		<item-type>Switch</item-type>
+		<label>Start collect the dust at the station</label>
+	</channel-type>
+	<channel-type id="cleanmopstart" advanced="true">
```
See above:
```suggestion
	<channel-type id="clean_mop_start" advanced="true">
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -299,6 +315,18 @@
+	</channel-type>
+	<channel-type id="cleanmopstart" advanced="true">
+		<item-type>Switch</item-type>
+		<label>Start mop wash</label>
```
```suggestion
		<label>Start Mop Wash</label>
```

truidix replied:
Thanks I need to get used to the writings you are using here.. :-)

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -299,6 +315,18 @@
+		<item-type>Switch</item-type>
+		<label>Start mop wash</label>
+	</channel-type>
+	<channel-type id="cleanmopstop" advanced="true">
```
See above:
```suggestion
	<channel-type id="clean_mop_stop" advanced="true">
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -299,6 +315,18 @@
+	</channel-type>
+	<channel-type id="cleanmopstop" advanced="true">
+		<item-type>Switch</item-type>
+		<label>Stop mop wash</label>
```
```suggestion
		<label>Stop Mop Wash</label>
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -183,6 +188,17 @@
+	<channel-type id="mop_drying_time">
+		<item-type>Number</item-type>
+		<label>Mop Drying Time</label>
+		<state pattern="%.0f %unit%" readOnly="true"/>
```
The channel won't have a unit unless it's defined as UoM. Since you are already updating it correctly with QuantityType with unit, you could use `Number:Time` instead.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -183,6 +188,17 @@
+	</channel-type>
+	<channel-type id="is_mop_drying">
+		<item-type>Switch</item-type>
+		<label>Mop cleaning active</label>
```
See below:
```suggestion
		<label>Mop Cleaning Active</label>
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -34,6 +34,9 @@
 			<channel id="fan" typeId="fan"/>
 			<channel id="vacuum" typeId="vacuum"/>
 			<channel id="segment" typeId="segment"/>
+			<channel id="collectdust" typeId="collectdust"/>
+			<channel id="cleanmopstart" typeId="cleanmopstart"/>
+			<channel id="cleanmopstop" typeId="cleanmopstop"/>
```
See below:
```suggestion
			<channel id="collect_dust" typeId="collect_dust"/>
			<channel id="clean_mop_start" typeId="clean_mop_start"/>
			<channel id="clean_mop_stop" typeId="clean_mop_stop"/>
```

truidix replied:
Yeah, thanks same here: I would better had have a look on how you writing stuff in here

jlaur commented on bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java:
```diff
@@ -370,6 +392,22 @@ private boolean updateVacuumStatus(JsonObject statusData) {
+        }
+        if (deviceCapabilities.containsKey(RobotCababilities.MOP_DRYING_REMAING_TIME)) {
+            updateState(CHANNEL_MOP_TOTALDRYTIME,
+                    new QuantityType<>(TimeUnit.SECONDS.toMinutes(statusInfo.getMopDryTime()), Units.SECOND));
```
Since just converted from seconds to minutes:
```suggestion
                    new QuantityType<>(TimeUnit.SECONDS.toMinutes(statusInfo.getMopDryTime()), Units.MINUTE));
```

or alternatively, to keep as seconds:
```suggestion
                    new QuantityType<>(statusInfo.getMopDryTime(), Units.SECOND));
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -183,6 +188,17 @@
 		<state pattern="%.0f" readOnly="true"/>
 	</channel-type>
+	<channel-type id="mop_drying_time">
+		<item-type>Number</item-type>
```
See below:
```suggestion
		<item-type>Number:Time</item-type>
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/robot/StatusDTO.java:
```diff
@@ -85,6 +85,12 @@ public class StatusDTO {
     private Integer mopForbiddenEnable;
+    @SerializedName("dry_status")
+    @Expose
+    private Integer isMapDryingActive;
```
Typo?
```suggestion
    private Integer isMopDryingActive;
```

truidix replied:
In deed: Typo: I'm going to fix that, too.

Thanks

jlaur reviewed (changes requested):
Thanks for the contribution! I have provided some minor feedback.

Also, since channels were added, you need to provide update instructions for the new channels. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types

truidix commented:
Hey  @jlaur , , thanks for your feedback and you patience.  I changed almost everything you suggested. 
Also I did not now about the update XML mechanism. Thanks for the hint. Even if I was wondering, that such a file does not exists I created a new one. Hopefully in the right way and it can be used.

truidix reviewed (commented):
I had a look on everything and did a few changes. Thanks for your patience!

jlaur commented on bundles/org.openhab.binding.miio/README.base.md:
```diff
@@ -252,6 +252,13 @@ Additionally depending on the capabilities of your robot vacuum other channels m
 | Switch  | status#mop_forbidden_enable       | Mop Forbidden              |
 | Switch  | status#is_locating                | Robot is locating          |
 | Number  | actions#segment                   | Room Clean  (enter room #) |
+| Switch  | actions#collectdust               | Start collect the dust at  |
+| Switch  | actions#cleanmopstart             | Start mop wash             |
+| Switch  | actions#cleanmopstop              | Stop mop wash              |
```
Should be updated after channels were renamed.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/update/vacuumchannels.xml:
```diff
@@ -0,0 +1,31 @@
+		<instruction-set targetVersion="1">
+			<add-channel id="collect_dust">
+				<type>miio:collect_dust</type>
+				<label>Start Collect Dust</label>
```
Label should be omitted when channel doesn't override channel type label.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/update/vacuumchannels.xml:
```diff
@@ -0,0 +1,31 @@
+			</add-channel>
+			<add-channel id="clean_mop_start">
+				<type>miio:clean_mop_start</type>
+				<label>Start Mop Wash</label>
```
Label should be omitted when channel doesn't override channel type label.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/update/vacuumchannels.xml:
```diff
@@ -0,0 +1,31 @@
+			</add-channel>
+			<add-channel id="clean_mop_stop">
+				<type>miio:clean_mop_stop</type>
+				<label>Stop Mop Wash</label>
```
Label should be omitted when channel doesn't override channel type label.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/update/vacuumchannels.xml:
```diff
@@ -0,0 +1,31 @@
+			</add-channel>
+			<add-channel id="mop_drying_time">
+				<type>miio:mop_drying_time</type>
+				<label>Mop Drying Time</label>
```
Label should be omitted when channel doesn't override channel type label.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/update/vacuumchannels.xml:
```diff
@@ -0,0 +1,31 @@
+			</add-channel>
+			<add-channel id="is_mop_drying">
+				<type>miio:is_mop_drying</type>
+				<label>Mop Cleaning Active</label>
```
Label should be omitted when channel doesn't override channel type label.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
For the update instructions to work, you also need to set the `thingTypeVersion` property. After that, please test that the migration works by creating a managed thing with a version before these changes, then update to your new version and verify that the new channels are automatically added.

truidix replied:
Hey @jlaur,

I had a look at the avm binding and the netatmo binding, because for me it looks like this two are the only on using the thingTypeVersion so far.  Both are using different approaches;

While netatmo is having own classes for the type and adding the thingTypeVersion into that classes, the avm binding is adding that property to the thing via the "update xml".  Id did the last one for miio: I added the version for the vacuum thing.
I also had a test and the channels added automatically. But on my test before without the "update xml" that happened, too. However, I'm not sure if I did the first tests on 3.4.  :-)

jlaur replied:
> I had a look at the avm binding and the netatmo binding, because for me it looks like this two are the only on using the thingTypeVersion so far.

There are plenty of bindings using it so far, so you might need to refine your search. 😉

> Id did the last one for miio: I added the version for the vacuum thing.

This static approach is fine and is also the most usual one.

> I also had a test and the channels added automatically. But on my test before without the "update xml" that happened, too.

Something must be wrong with that test then. Did you create the thing in the UI before performing the migration? Without update instructions or a `thingTypeVersion` triggering the instructions, the channels will not be added automatically - only when recreating the thing.

jlaur reviewed (changes requested):
A few additional comments.

marcelrv commented:
@jlaur the binding does part of this dynamically (dependent on the features reported by the robot). 
Don't know if that plays a role.

note: the vacuum thing in miio handles these optional channels different from the miio:basic. In miio:basic all channels are created dynamically, but in miio:vacuum, the definitions are in the xml and depending on the features of the robot these channels are connected

jlaur commented:
> the binding does part of this dynamically (dependent on the features reported by the robot).
> Don't know if that plays a role.

Ah, thanks - it does. If channels are **created** dynamically, we don't need update instructions. If channels are **created** statically but **removed** dynamically (depending on features etc.), we still need instructions. So maybe it worked because the channels are added dynamically by the binding?

truidix commented:
@marcelrv,I got some DCO warning, I'm not sure where the come from. I missed it on  a few commits, but fixed it on mine (I missed the configuration in on my other computer  ). I looks the commits, where DCO fails are not mine.
Unfortunately, I synchronized my fork before. May these DCO errors come from that sync?


Can/should I revert "the "update xml" changes?

jlaur commented on bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/robot/RobotCababilities.java:
```diff
@@ -38,7 +38,14 @@ public enum RobotCababilities {
+    CLEAN_MOP_START("dry_status", "actions#cleanmopstart", "miio:cleanmopstart", ""),
+    CLEAN_MOP_STOP("dry_status", "actions#cleanmopstop", "miio:cleanmopstop", ""),
+    MOP_DRYING("dry_status", "status#is_mop_drying", "miio:is_mop_drying", ""),
+    MOP_DRYING_REMAING_TIME("dry_status", "status#mop_drying_time", "miio:mop_drying_time", ""),
```
```suggestion
    MOP_DRYING_REMAINING_TIME("dry_status", "status#mop_drying_time", "miio:mop_drying_time", ""),
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/update/instructions.xml:
The last two added channels are missing here.

truidix replied:
Hey, yes I did not add them because I'm not sure if we really need them. As  @marcelrv  said the some channels are added dynamically, whenever the vacuum cleaner is having a specific capability. So I don’t know if we really need that xml upgrade file.  In my opinion we don’t.

jlaur commented on bundles/org.openhab.binding.miio/README.base.md:
```diff
@@ -252,6 +252,14 @@ Additionally depending on the capabilities of your robot vacuum other channels m
+| Switch  | actions#collect_dust              | Start collect the dust at  |
+| Switch  | actions#clean_mop_star            | Start mop wash             |
+| Switch  | actions#clean_mop_stop            | Stop mop wash              |
+| Number  | status#mop_drying_time            | Mop Drying Time            |
```
Not sure, but casing seems to be inconsistent here:
```suggestion
| Number  | status#mop_drying_time            | Mop drying time            |
```

jlaur commented on bundles/org.openhab.binding.miio/README.base.md:
```diff
@@ -252,6 +252,14 @@ Additionally depending on the capabilities of your robot vacuum other channels m
+| Switch  | actions#clean_mop_stop            | Stop mop wash              |
+| Number  | status#mop_drying_time            | Mop Drying Time            |
+| Switch  | status#is_mop_drying              | Mop cleaning active        |
+| Number  | status#dock_state_id              | Dock Status id             |
+| String  | status#dock_state                 | Dock Status message        |
```
Depending on decided casing:
```suggestion
| Number  | status#dock_state_id              | Dock status id             |
| String  | status#dock_state                 | Dock status message        |
```

truidix replied:
I changed them to lower case, even if some other descriptions using upper case.

jlaur commented on bundles/org.openhab.binding.miio/README.md:
```diff
@@ -646,6 +646,13 @@ Additionally depending on the capabilities of your robot vacuum other channels m
 | Switch  | status#mop_forbidden_enable       | Mop Forbidden              |
 | Switch  | status#is_locating                | Robot is locating          |
 | Number  | actions#segment                   | Room Clean  (enter room #) |
+| Switch  | actions#collect_dust              | Start collect the dust at  |
```
See above.
```suggestion
| Switch  | actions#collect_dust              | Start collect the dust     |
```

jlaur commented on bundles/org.openhab.binding.miio/README.md:
```diff
@@ -646,6 +646,13 @@ Additionally depending on the capabilities of your robot vacuum other channels m
+| Switch  | actions#collect_dust              | Start collect the dust at  |
+| Switch  | actions#clean_mop_star            | Start mop wash             |
+| Switch  | actions#clean_mop_stop            | Stop mop wash              |
+| Number  | status#mop_drying_time            | Mop Drying Time            |
```
See above:
```suggestion
| Number  | status#mop_drying_time            | Mop drying time            |
```

jlaur commented on bundles/org.openhab.binding.miio/README.md:
```diff
@@ -646,6 +646,13 @@ Additionally depending on the capabilities of your robot vacuum other channels m
+| Switch  | actions#clean_mop_stop            | Stop mop wash              |
+| Number  | status#mop_drying_time            | Mop Drying Time            |
+| Switch  | status#is_mop_drying              | Mop cleaning active        |
+
```
Did you forget to add dock status channels here?

truidix replied:
Yeah i missed to run the Readmehelper,.. :-(

jlaur reviewed (changes requested):
A few additional comments after last commit.

jlaur commented:
> note: the vacuum thing in miio handles these optional channels different from the miio:basic. In miio:basic all channels are created dynamically, but in miio:vacuum, the definitions are in the xml and depending on the features of the robot these channels are connected

I'm confused by this statement. The channels in this PR are for `miio:vacuum`, so according to you they are statically defined. Therefore I would expect that update instructions are needed. But @truidix wrote that they appeared automatically even without instructions?

jlaur commented:
> depending on some capabilities (most of them are checked via "get_status") some channels are added dynamically. Therefore, I would say that the xml file makes no sense, as the channels I added are all dynamic, I tested it without this xml file, and all channels appeared, as they should. I also gave the snapshot jar to some other people and did not get any negative feedback from them.

Okay, thanks, so I think I get it now. In that case you can remove the update instructions as well as the channels from the thing type definition in `vacuumThing.xml`. It looks like the channels will be dynamically created here based on your defined `RobotCapabilities`:
https://github.com/openhab/openhab-addons/blob/888f4388a31b38ba3214c3920484b2fc4ea97f89/bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/handler/MiIoVacuumHandler.java#L740-L777

truidix commented:
Hey @jlaur, thanks for the feedback.
I removed the "update xml". I don’t know yet the whole procces/startucre of openhab binding, but I would say the definition at the vacuumThing.xml is still needed. I had a look into it and found some form the previous defined optional channels in it. 
For testing I also removed one of my new channels from the xml, in result that channel is not available anymore. 
So I’m pretty sure  that the entries must be retained.

jlaur commented:
> I removed the "update xml". I don’t know yet the whole procces/startucre of openhab binding, but I would say the definition at the vacuumThing.xml is still needed. I had a look into it and found some form the previous defined optional channels in it.
> For testing I also removed one of my new channels from the xml, in result that channel is not available anymore.
> So I’m pretty sure that the entries must be retained.

@marcelrv - can you comment on this? From my understanding the channels are created dynamically according to capabilities. In this case they shouldn't be statically declared in the XML as well (only the channel types of course, but not the channels)? If they are not created dynamically but only statically, the update instructions would still be needed. But this would contradict your previous statement that vacuum channels are created dynamically, and it also seems strange that they appeared anyway for @truidix after the upgrade. So something is still fishy...

@truidix - when you removed the channels from the XML and lost them, did you remove only the channels, or also the channel types? Just to be sure...

truidix commented:
hey @jlaur ,

you are right, I removed both. I tried it again with just removing the channels and kept the types. 
That looks great! I do have a s5 without that dockingstation state and mop features I tried it with them, too. In addition, it works like it should. I also missed some renaming: I fixed that too.  :-) Thanks

marcelrv commented:
> @marcelrv - can you comment on this? From my understanding the channels are created dynamically according to capabilities. In this case they shouldn't be statically declared in the XML as well (only the channel types of course, but not the channels)? If they are not created dynamically but only statically, the update instructions would still be needed. But this would contradict your previous statement that vacuum channels are created dynamically, and it also seems strange that they appeared anyway for @truidix after the upgrade. So something is still fishy...

Hallo @jlaur 
I think you already saw the relevant piece of code.
So the channel**Types** are not created dynamically, they are created statically hence they are defined in the XML
At runtime, depending on the robot capabilities a channel of this channeltype is created and linked to the thing.

NB. The above is only for channels that are not available for all robots.
**Channels that are available for each robot are statically defined in the XML.** 

This latter type of channels is not in play for this specific PR as it only deals with these ' dynamic' linked channels
Hope this clarifies.

@jlaur Can you advice if the `update xml` is needed in the case. 
Is it needed to make the framework aware of a new channeltype, or is the framework properly 'aware'  when the binding creates the channel dynamically.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -215,6 +218,41 @@
+	</channel-type>
+	<channel-type id="clean_mop_stop">
+		<item-type>Switch</item-type>
+		<label>Stop mop wash</label>
```
```suggestion
		<label>Stop Mop Wash</label>
```

truidix replied:
thanks.. i changed that.

jlaur replied:
Only half. 😉

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/i18n/miio.properties:
Please regenerate this file once the last label has been fixed.

jlaur replied:
Still not regenerated.

jlaur replied:
See https://www.openhab.org/docs/developer/utils/i18n.html#generating-i18n-properties-file

truidix replied:
Thanks for the hint: it is now regenerated.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -18,8 +18,10 @@
+
 		<properties>
 			<property name="vendor">Xiaomi</property>
+			<property name="thingTypeVersion">1</property>
```
Please also remove this again.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -184,6 +186,7 @@
 		<state pattern="%.0f" readOnly="true"/>
 	</channel-type>
 
+
```
```suggestion
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -300,6 +338,7 @@
 		<label>Vacuum Room [room#]</label>
 	</channel-type>
 
+
```
```suggestion
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -215,6 +218,41 @@
+		<item-type>Switch</item-type>
+		<label>Stop mop wash</label>
+	</channel-type>
+
```
```suggestion
```

jlaur reviewed (changes requested):
A few last comments.

jlaur commented:
> Can you advice if the `update xml` is needed in the case.
> Is it needed to make the framework aware of a new channeltype, or is the framework properly 'aware' when the binding creates the channel dynamically.

This is not needed, the update instructions are only needed to update the redundant channel information within jsondb for managed things.

maniac103 commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -215,6 +218,41 @@
+	</channel-type>
+	<channel-type id="collect_dust">
+		<item-type>Switch</item-type>
+		<label>Start Collect Dust</label>
```
`Start Dust Collection` or `Start Collecting Dust`? Current wording doesn't sound like valid English.

truidix commented:
@jlaur , @maniac103 ,I changed the few things.

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -215,7 +215,41 @@
+	</channel-type>
+	<channel-type id="clean_mop_stop">
+		<item-type>Switch</item-type>
+		<label>Stop Mop wash</label>
```
```suggestion
		<label>Stop Mop Wash</label>
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -215,7 +215,41 @@
+	</channel-type>
+	<channel-type id="collect_dust">
+		<item-type>Switch</item-type>
+		<label>Start collecting dust</label>
```
```suggestion
		<label>Start Collecting Dust</label>
```

truidix replied:
I'm really struggling, when you want words in first letter upper case..  :-(

jlaur reviewed (changes requested):
Still some label tweaking needed. 🙂

jlaur commented on bundles/org.openhab.binding.miio/README.base.md:
```diff
@@ -252,6 +252,13 @@ Additionally depending on the capabilities of your robot vacuum other channels m
 | Switch  | status#mop_forbidden_enable       | Mop Forbidden              |
 | Switch  | status#is_locating                | Robot is locating          |
 | Number  | actions#segment                   | Room Clean  (enter room #) |
+| Switch  | actions#collect_dust              | Start collect the dust     |
```
```suggestion
| Switch  | actions#collect_dust              | Start collecting dust      |
```

jlaur commented on bundles/org.openhab.binding.miio/README.md:
```diff
@@ -646,6 +646,13 @@ Additionally depending on the capabilities of your robot vacuum other channels m
 | Switch  | status#mop_forbidden_enable       | Mop Forbidden              |
 | Switch  | status#is_locating                | Robot is locating          |
 | Number  | actions#segment                   | Room Clean  (enter room #) |
+| Switch  | actions#collect_dust              | Start collect the dust     |
```
```suggestion
| Switch  | actions#collect_dust              | Start collecting dust      |
```

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/thing/vacuumThing.xml:
```diff
@@ -19,7 +19,7 @@
 
 		<properties>
-			<property name="vendor">Xiaomi</property>
+			<property name="vendor">Xiaomi</property>			
```
It seems like some spaces are appended here:
```suggestion
			<property name="vendor">Xiaomi</property>
```

truidix commented:
@jlau,I applied spotless and changed the requested wording.. I’m really fighting on when you want to write every word wit first letter upper case.. :-( Hopefully everything is fine now and it can be merged..

jlaur commented:
> I applied spotless and changed the requested wording.. I’m really fighting on when you want to write every word wit first letter upper case.. :-( Hopefully everything is fine now and it can be merged..

See https://github.com/openhab/openhab-addons/pull/15704#discussion_r1374956803 regarding the casing.

Last thing remaining is the I18N properties file, see https://github.com/openhab/openhab-addons/pull/15704#discussion_r1383855383

marcelrv commented:
Can you also once more review if the readme & readme.base needs to be adjusted to reflect all the changes in the channel naming

jlaur commented on bundles/org.openhab.binding.miio/src/main/resources/OH-INF/i18n/miio.properties:
```diff
@@ -155,7 +181,14 @@ channel-type.miio.water_box_carriage_status.label = Water Box Carriage State
 channel-type.miio.water_box_status.label = Water Box State
 
-# Thing status descriptions
+# channel types
+
+channel-type.miio.cleanmopstart.label = Start mop wash
+channel-type.miio.cleanmopstop.label = Stop mop wash
+channel-type.miio.collectdust.label = Start collecting dust at the station
+
```
These needs to be manually removed since they don't exist anymore. The tool will only move them, but not delete them, since it can't know for sure if they are referenced dynamically.

truidix replied:
@jlaur  I removed them

truidix commented:
> Can you also once more review if the readme & readme.base needs to be adjusted to reflect all the changes in the channel naming

Hey @marcelrv , I had another look into it and double checked the vaccumThing typeids with the README.base (I also regenerated the README). It looks god for me, now.

jlaur reviewed (approved):
LGTM

jlaur commented:
@marcelrv - I will give you a chance to have a final look and approve. I would then plan to merge in the evening, unless you would approve or request more time before then.

marcelrv reviewed (changes requested):
can you make these  public static final String  's in MiiIoBindingConstats.java
For the rest it looks good to me

truidix commented:
Hey @marcelrv,

as requested I moved the two channels to MiIoBindingConstants.

marcelrv commented:
Lgtm
